Source/Core/InputCommon/ControllerInterface/ControllerInterface.cpp

```diff
@@ -92,6 +94,24 @@ void ControllerInterface::Initialize(const WindowSystemInfo& wsi)
 
   if (m_populating_devices_counter.fetch_sub(1) == 1 && !devices_empty)
     InvokeDevicesChangedCallbacks();
+
+  // Given that a lot of devices population calls are async, wait for the first device to be added
+  // so that we can then safely load the default input configs for different controllers,
+  // as otherwise their default device would end up being empty and the default key bindings would
+  // not be linked to any device (by design, the first device added to the CI is the default default
+  // device, and it's usually the "keyboard and mouse" on every OS, which should always be present).
+  if (wait_for_first_device)
+  {
+    // Somebody might have triggered InvokeDevicesChangedCallbacks() without actually
+    // adding any device, so check against that and reset the event.
+    // In some extremely rare occasions devices might change in between the two checks,
+    // but at worse we'll wait a few seconds.
```

Comment:
It should be possible to fix this race condition properly. Waiting for five seconds sounds rather annoying.

A `Common::Event` is essentially a wrapper around a mutex and a condition variable. The key to avoiding the race condition is to take the mutex before you check `m_devices.empty()` and keep holding it until you're done with the `wait_for` call. I don't think it's possible to do this using the `Common::Event` wrapper class, so my suggestion is that you use a raw `std::condition_variable` in the following way:

1. Lock `m_devices_mutex`
2. Check `m_devices.empty()` (do it anew – don't just read the old `devices_empty` variable, since we've unlocked `m_devices_mutex` since then)
3. If it was empty, call `wait_for` on the condition variable
4. Unlock `m_devices_mutex`

This would also require that the code that calls notify on the condition variable has `m_devices_mutex` locked. I suppose it could make sense to put the notify call in `AddDevice` instead of in `InvokeDevicesChangedCallbacks`?

Reply:
Should be all safe now.